### DISCUSSION TITLE
Trying to get assertEmitted to work with a model

### DISCUSSION THREAD

**Discussion by telkins** [livewire/livewire#3792] [1. Help] [answered]
Hi,

For certain conditions, like when a particular model is updated, I fire an event to indicate as much...like so:
```php
$this->emit('my-model-updated', $myUpdatedModel);
```

When I test, I use the following assertion:
```php
   ->assertEmitted('my-model-updated', $myModelIExpectToBeUpdated);
```

It's failing.  😳

I'm dumping `$myUpdatedModel` just before emitting the event and all of its attributes match the attributes listed in the assertion failure message in the console:
```bash
Failed asserting that an event [my-model-updated] was fired with parameters: [{"attribute_key":"attribute_value",...}]
```

I'm not sure why this is happening, obviously, and would welcome any help/input.

I wondered if it had anything to do with how the model might be serialized.  The timestamps, for example, both show the same date/time, but in the error message it shows what appears to be the model's date/time casting (with microseconds, for example).

I also wondered whether or not the model passed in `emit()` and the model passed in `assertEmitted()` were "in sync"....so I added `->fresh()` to each to try to make sure they were both "fresh" from the DB.

Finally, I asked Uncle Google and got no help...so here I am.  🤓 

Thanks in advance for any help.  🚀 

Travis

**Comment by joshhanley** [comment]
@telkins yeah it's probably a serialisation issue.

I personally only pass the model id/uuid into into the emit function. 

Hope this helps!

**Reply by telkins** [reply]
@joshhanley Thx.  I'm guessing this isn't something worth worrying about?

**Reply by joshhanley** [reply]
@telkins probably not, from my perspective.

I've always found directly comparing models in tests can be problematic.

That's why if you look at Laravel's `->is` method on a model, you will see it only compares ID, model and database connection (I think?).

**Comment by ralphjsmit** [comment, marked as answer]
After some digging into the GitHub issues here, I found a small remark from Caleb that you could also pass a callback to the `->assertEmitted()` function. When looking into the source code, I found that this callback would receive the `$event` and `$params` parameters.

Checkout this example below where I test whether there are three variables emitted with the event. It's pretty self-explanatory:
```php
        $product = \App\Models\Product::factory()->create();

        $livewire = Livewire::test(Product::class, [ 'product' => $product ])
            ->assertSee($product->name);

        $livewire->set('amount', 3)
            ->call('addToCart')
            ->assertEmitted('addToCart', function($event, $params) use ($product) {
                return $params[0]['product']->is($product);
            }, $product->getOptionsAsAmount(), 3);
```

PS: if you dump the `$params` variable, you can get a good idea about what it looks like.

**Reply by jacksontong** [reply]
this should mark as answer.
